aten/src/ATen/VmapModeRegistrations.cpp

```diff
@@ -0,0 +1,114 @@
+#include <torch/library.h>
+#include <ATen/core/boxing/KernelFunction.h>
+
+using torch::CppFunction;
+
+namespace at {
+
+// Note: [DispatchKey::VmapMode usage]
+// Whenever we're inside a vmap, all Tensors dispatch on this key. At the moment,
+// this key is used to disable random operations inside of vmap. If you are looking
+// for Batching Rules, those are registered with DispatchKey::Batched instead.
+//
+// Note: [Ambiguity of random operations inside vmap]
+// Random operations have an ambiguity where it isn't clear if they should
+// apply the same randomness or apply different randomness. For example:
+//
+// >>> vmap(lambda t: torch.rand(1))(torch.zeros(5))
+// Should the above return the same random number 5 times, or a different one?
+//
+// We haven't made a decision on that yet so we are temporarily banning random
+// operations inside of vmap while we gather user feedback.
+
+template <typename... Args> Tensor unsupportedRandomOp(Args... args) {
+  TORCH_CHECK(false, "vmap: We do not yet support calling random operations inside of vmap. ",
+              "Please perform random operations outside of vmap as a workaround");
+}
+
+template <typename... Args> Tensor& unsupportedRandomOp_(Args... args) {
+  TORCH_CHECK(false, "vmap: We do not yet support calling random operations inside of vmap. ",
+              "Please perform random operations outside of vmap as a workaround");
+}
+
+TORCH_LIBRARY_IMPL(_, VmapMode, m) {
+  m.fallback(torch::CppFunction::makeFallthrough());
+}
+
+TORCH_LIBRARY_IMPL(aten, VmapMode, m) {
+  // NB: I'd really like to register a special kernel like
+  // CppFunction::makeNamedNotSupported() to avoid listing out the types of everything.
+  // However, registering e.g. CppFunction::makeNamedNotSupported() as an implementation
+  // only works for operators that support boxing.
+#define TENSOROPTIONS c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>
```

Comment:
cc @smessmer @bhosmer @gchanan Downside of exploded tensor options

Reply:
I really wanted to do an e.g. `decltype(native::randn)` here to avoid listing out the types. However there were two problems with that:
1) we've got a lot of overloads
2) native::randn takes `const TensorOptions&`, so a decltype wouldn't give the correct thing